subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppIncrementalCompileIntegrationTest.groovy

```diff
@@ -80,6 +83,100 @@ class CppIncrementalCompileIntegrationTest extends AbstractInstalledToolChainInt
         sharedLibrary("build/lib/main/debug/hello").assertExists()
     }
 
+    def "removes stale installed executable and library file when all source files for executable are removed"() {
+        settingsFile << "include 'app', 'greeter'"
+        def app = new IncrementalCppStaleLinkOutputAppWithLib()
+
+        given:
+        buildFile << """
+            project(':app') {
+                apply plugin: 'cpp-executable'
+                dependencies {
+                    implementation project(':greeter')
+                }
+            }
+            project(':greeter') {
+                apply plugin: 'cpp-library'
+            }
+"""
+        app.library.writeToProject(file("greeter"))
+        app.executable.writeToProject(file("app"))
+
+        and:
+        succeeds "assemble"
+        app.library.applyChangesToProject(file('greeter'))
+        app.executable.applyChangesToProject(file('app'))
+
+        expect:
+        succeeds "assemble"
+        result.assertTasksExecuted(":greeter:compileDebugCpp", ":greeter:linkDebug", ":greeter:assemble",
+            ":app:compileDebugCpp", ":app:linkDebug", ":app:installDebug", ":app:assemble",
+            ":assemble")
+        result.assertTasksNotSkipped(":app:compileDebugCpp", ":app:linkDebug", ":app:installDebug", ":app:assemble")
+        result.assertTasksSkipped(":assemble", ":greeter:compileDebugCpp", ":greeter:linkDebug", ":greeter:assemble")
+
+        executable("app/build/exe/main/debug/app").assertDoesNotExist()
```

Comment:
These checks will stop checking anything if we change the output locations. We can make the checks more robust by asserting that the files exist first, or by asserting something about the contents of the build directory as a whole.

Reply:
I added robustness by asserting the files exist first.